WIP feat(oauth): signal action=signup or action=signin at end of oauth flow.

rfk opened the pull request:
Here's my WIP experiment with signalling "action=signup" or "action=signin" at the end of the OAuth flow.  Most definitely now ready yet, but thought I'd put it out there early.

Fixes #2345

shane-tomlinson commented on app/scripts/models/auth_brokers/redirect.js:
```diff
@@ -17,14 +17,22 @@ define([
             // really, we should be parsing the URL and adding the error
             // parameter. That requires more code than this.
-            var separator = redirectTo.indexOf('?') === -1 ? '?' : '&';
+            separator = redirectTo.indexOf('?') === -1 ? '?' : '&';
```
Maybe it's time to parse the URI, then make use of [lib.js->objToSearchString](https://github.com/mozilla/fxa-content-server/blob/4898dc006863fe26746b7562cc5dad78c36ca2f5/app/scripts/lib/url.js#L49), this is starting to look a bit out of control. Not sure how much more effort that would be.

rfk replied:
yes, it's definitely time

shane-tomlinson commented on app/scripts/models/auth_brokers/web-channel.js:
```diff
@@ -103,13 +103,13 @@ define([
       // event handlers before the flow completes.
       var self = this;
-      return p().delay(100).then(_.bind(self.finishOAuthFlow, self, account));
+      return p().delay(100).then(_.bind(self.finishOAuthFlow, self, account, { action: 'signup' }));
```
nit: this line wraps in GitHub side by side view, can you split it before the `.delay(100)` or before the `.then`?

shane-tomlinson commented on app/scripts/models/auth_brokers/oauth.js:
```diff
@@ -119,6 +119,8 @@ define([
 
     afterSignIn: function (account, additionalResultData) {
+      // Signal to the RP that this was an existing account sign-in.
+      additionalResultData.action = 'signin';
```
These magic strings should probably live in [lib/constants.js](https://github.com/mozilla/fxa-content-server/blob/4898dc006863fe26746b7562cc5dad78c36ca2f5/app/scripts/lib/constants.js)

shane-tomlinson commented:
@rfk - can you rebase this and I can review when ready?

pdehaan commented on app/tests/spec/models/auth_brokers/oauth.js:
```diff
@@ -156,9 +163,12 @@ function (chai, sinon, Session, p, OAuthClient, Assertion, AuthErrors,
         return broker.afterResetPasswordConfirmationPoll(account)
           .then(function () {
-            assert.isTrue(broker.finishOAuthFlow.calledWith(account));
+            assert.isTrue(broker.finishOAuthFlow.calledWith(account, {;
```
What's this little `;` doing at the end here?

rfk commented:
OK, rebased and cleaned up a little.  I'm not thrilled about this architecturally, it seems to spread the concern across too much of the codebase.  But I don't have a better suggestion, since it really is a property of the flow rather than a property of any particular object in the system.

I toyed with putting some sort of "isNewAccount()" indicator on the account object, but it didn't seem to come out any cleaner by the time you figure out how to un-set this flag properly after the first sign-in.  At least here the logic is contained to the oauth backends.

shane-tomlinson commented on app/scripts/models/auth_brokers/redirect.js:
```diff
@@ -59,14 +68,15 @@ define([
     },
 
-    finishOAuthFlow: function (account) {
+    finishOAuthFlow: function (account, additionalResultData) {
```
Maybe instead of only using `finishOAuthFlow`, there could be two functions `finishOAuthSignUp` and `finishOAuthSignIn`. These two functions would each call `finishOAuthFlow` with the appropriate `action`.

```js
return self.finishOAuthFlow(account, {
  action: Constants.OAUTH_ACTION_SIGNIN
}).then(function () {
```

This could be converted to:
```js
return self.finishOAuthSignIn(account)
  .then(function () {
```

rfk replied:
That is *substantially* cleaner; nice one @shane-tomlinson!

shane-tomlinson commented on app/scripts/models/auth_brokers/redirect.js:
```diff
@@ -19,15 +20,23 @@ define([
           var redirectTo = result.redirect;
+          var params = {};
+
+          var startOfParams = redirectTo.lastIndexOf('?');
```
Maybe extract the URL->search params section into lib/url.js? It certainly could be reused in the future, and it'll hopefully reduce the amount of code in this function.

shane-tomlinson commented on app/scripts/models/auth_brokers/oauth.js:
```diff
@@ -92,6 +95,18 @@ define([
     },
 
+    finishOAuthSignInFlow: function (account, additionalResultData) {
+      if (! additionalResultData) { additionalResultData = {}; }
```
This is my lone nit in this PR. 

We usually set defaults/missing parameters using this form: 
```js
additionalResultData = additionalResultData || {};
```

shane-tomlinson commented:
I manually merged this in ca3d564fb7b967afc7b4687dd3802c159e08dc7b. Thanks @rfk!
